tests/queries/0_stateless/02845_threads_count_in_distributed_queries.sql.j2

```diff
@@ -0,0 +1,206 @@
+SETTINGS parts_to_delay_insert=0;
+
+CREATE TABLE test_lot_of_parts_distributed
+(
+    `a` String,
+    `b` LowCardinality(String),
+    `c` DateTime64(3),
+    `val` String,
+)
+ENGINE = Distributed(test_shard_localhost, currentDatabase(), 'test_lot_of_parts', rand());
+
+-- we need a lot of parts to make sure that we will have a lot of streams
+SYSTEM STOP MERGES test_lot_of_parts;
+INSERT INTO test_lot_of_parts (a, b, c, val)
+   SELECT
+     'foo' as a,
+     'bar' as b,
+     _CAST('1683504000', 'DateTime64') as c,
+     'baz' as val
+   FROM numbers_mt(500)
+   SETTINGS max_block_size = 1, min_insert_block_size_bytes=1, min_insert_block_size_rows=1; --every row will be in separate part
+
+select count() from system.parts where table = 'test_lot_of_parts' and active;
+
+-- query which uses optimize_read_in_order=1
+SELECT DISTINCT
+    'val' AS fieldType,
+    val AS value
+FROM test_lot_of_parts_distributed
+WHERE a = 'foo' AND value != '' AND positionCaseInsensitiveUTF8(value, 'baz') > 0 AND b = 'bar' AND c >= _CAST('1683504000', 'DateTime64')
+ORDER BY c DESC
+LIMIT 5
+SETTINGS log_comment='check_concurrency_in_remote_queries4' FORMAT Null;
+
+SYSTEM FLUSH LOGS;
+
+-- it seem like the number of threads here reported here is still higher than expected
+-- max_threads=1 => peak_threads=7
+-- max_threads=2 => peak_threads=14
+-- max_threads=5 => peak_threads=25
+-- may be it's a matter of the way how we count threads in thread_log, or there is another problem here
+-- anyway it still much better than before (see the next test in that file - it uses 500 threads).
+
+WITH
+    maxIntersections(
+        toUnixTimestamp64Micro(query_start_time_microseconds),
+        toUnixTimestamp64Micro(event_time_microseconds)
+    ) as peak_threads
+SELECT
+    peak_threads < toUInt64(getSetting('max_threads')) * 8
+FROM system.query_thread_log
+WHERE
+    event_time > now() - 60
+    AND query_id IN (
+        SELECT query_id
+        FROM system.query_log
+        WHERE
+            type = 'QueryFinish'
+            AND event_time > now() - 60
+            AND log_comment = 'check_concurrency_in_remote_queries4'
+            AND current_database = currentDatabase()
+    );
+
+
+-- that is actually a bad behaviour, but it used to work like that for a long time.
+-- now is happens only for async_socket_for_remote=0 (while it is 1 by default)
+SELECT DISTINCT
+    'val' AS fieldType,
+    val AS value
+FROM test_lot_of_parts_distributed
+WHERE a = 'foo' AND value != '' AND positionCaseInsensitiveUTF8(value, 'baz') > 0 AND b = 'bar' AND c >= _CAST('1683504000', 'DateTime64')
+ORDER BY c DESC
+LIMIT 5
+SETTINGS log_comment='check_concurrency_in_remote_queries5', async_socket_for_remote=0 FORMAT Null;
+
+SYSTEM FLUSH LOGS;
+
+WITH
+    maxIntersections(
+        toUnixTimestamp64Micro(query_start_time_microseconds),
+        toUnixTimestamp64Micro(event_time_microseconds)
+    ) as peak_threads
+SELECT
+    peak_threads >= 500 -- we have 500 parts
```

Comment:
check that example:
https://gist.github.com/filimonov/e4aeaa986123bde69a71d26e8be30f5d

Reply:
>check that example:
> [gist.github.com/filimonov/e4aeaa986123bde69a71d26e8be30f5d](https://gist.github.com/filimonov/e4aeaa986123bde69a71d26e8be30f5d)

Not sure what I should see there, but it reports `peak_threads=14`, so the test should fail

But anyway, now you can use `peak_threads_usage`, I guess? https://github.com/ClickHouse/ClickHouse/pull/54335

Reply:
>  it reports peak_threads=14 

yes. While max_threads is 5. 

> Not sure what I should see there

count of rows in system.query_thread_log is about 500. So simple `count() from system.query_thread_log` will say 500 instead of of 14. 

> But anyway, now you can use peak_threads_usage, I guess?

Yes, I can. Just wanted to stabilize it first before that.

Reply:
> now you can use peak_threads_usage, I guess?

I've decided not to do that in that PR (to make it possible to backport it), I can create a follow-up PR if you want (or keep it as is, seems like that calculation using query_thread_log works ok).